### PR TITLE
Allow @objc(Name) on enums

### PR DESCRIPTION
**Pull request by lilyball**
This implements support for

``` swift
@objc(ObjCEnum) enum SwiftEnum : Int {
    case One = 1, Two, Three
}
```

A new macro for the generated header is defined called `SWIFT_ENUM_NAMED` and the above declaration looks like

``` objc
typedef SWIFT_ENUM_NAMED(NSInteger, ObjCEnum, "SwiftEnum") {
    ObjCEnumOne = 1,
    ObjCEnumTwo = 2,
    ObjCEnumThree = 3,
};
```

Since `swift_name` doesn't actually work properly yet when importing Obj-C enums, there's an XFAILed test that tests the clang importer here.

This does NOT cover support for `@objc(Name)` on the individual `case` declarations.

### CODE REVIEW

File: test/ClangModules/enum-objc.swift [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// FIXME: @objc(Renamed) enums should work
+// XFAIL: *
+
+// RUN: %target-swift-frontend -emit-sil %s -import-objc-header %S/Inputs/enum-objc.h -verify
```
Comment (kballard) [inline test/ClangModules/enum-objc.swift]: I wasn't sure if this sort of test is the right way to do this, but I was using the `test/ClangModules/enum-new.swift` test as a prototype.

**Comment by jrose-apple** [comment]
Thanks for working on this!

File: test/ClangModules/enum-objc.swift [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// FIXME: @objc(Renamed) enums should work
```
Comment (DougGregor) [inline test/ClangModules/enum-objc.swift]: FWIW, you should be able to add the flag -enable-swift-name-lookup-tables to this test to get it working now. The Swift name lookup tables aren't on by default just yet, but they address this problem already.

**Comment by kballard** [comment]
I've updated the PR as per feedback. Tests are now part of the relevant commits, the enum constant prefix bit was pulled out into #743, and I've added relevant code to the resolve-cross-language test.

A couple of the tests have `// FIXME` comments for #743, because they use `__attribute__((swift_name(...)))` on enum constants in order to sidestep the enum constant prefix stripping. The tests will be unaffected by #743 but should ideally have the attributes removed once #743 lands.

**Comment by jrose-apple** [comment]
Looks great. Thanks, Kevin!

I merged #743 just now. Do you want to fix these tests first or just do a follow-up?

**Comment by kballard** [comment]
I'll fix the tests now.

**Comment by kballard** [comment]
@jrose-apple Rebased and updated to remove the FIXMEs.

**Comment by jrose-apple** [comment]
Aren't these commits in the wrong order? Ideally every commit should be build-and-testable on its own, although I suppose that matters less in a merge.

**Comment by kballard** [comment]
@jrose-apple GitHub bug, it's showing the commits in the wrong order. If you look at the branch they're ordered Sema, PrintObjC, ClangImporter.

**Comment by jrose-apple** [comment]
Weird, okay. Doing one more check on my machine, then I'll merge.

("It's not that I don't trust you…")

**Comment by kballard** [comment]
Yeah, it's annoying. I filed a support ticket over a year ago about this, and they still haven't fixed it :/

**Pull request merged**